core/src/sumeragi/main_loop.rs

```diff
@@ -421,8 +416,25 @@ fn commit_block<F: FaultInjection>(
         %block_hash, "Committing block"
     );
 
+    *current_topology = committed_block.header().committed_with_topology.clone();
+    current_topology.lift_up_peers(
+        &committed_block
+            .signatures()
+            .into_iter()
+            .map(|s| s.public_key().clone())
+            .collect::<Vec<PublicKey>>(),
+    );
```

Comment:
Isn't this approach has potential for abuse too? `ProxyTail` can manipulate whose signatures will go into the block and can control their orded.

Reply:
Yes, but it is less. Order is preserved. So the proxy tail is able to promote b set peers into the a set or demote a set peers into the b set. But is not able to reorder arbitrarely.